Load PostGIS raster using manual URI string. (#5385)
qgis/QGIS-Documentation: docs/pyqgis_developer_cookbook/loadlayer.rst

This creates a new layer and adds it to the current project (making it appear
in the layer list) in one step.

To load a PostGIS raster:

> PostGIS rasters, similar to PostGIS vectors, can be added to a project using a URI string. It is efficient to create a dictionary of strings for the database connection parameters. The dictionary is then loaded into an empty URI, before adding the raster. Note that None should be used when it is desired to leave the parameter blank: 

Reviewer: ```suggestion
0123456789012345678901234567890123456789012345678901234567890123456789
PostGIS rasters, similar to PostGIS vectors, can be added to a project using a URI string.
It is efficient to create a dictionary of strings for the database connection parameters.
The dictionary is then loaded into an empty URI, before adding the raster.
Note that None should be used when it is desired to leave the parameter blank: 
```

This creates a new layer and adds it to the current project (making it appear
in the layer list) in one step.

To load a PostGIS raster:

PostGIS rasters, similar to PostGIS vectors, can be added to a project using a URI string.
It is efficient to keep a reusable dictionary of strings for the database connection parameters. 
This makes it easy to edit the dictionary for the applicable connection.
The dictionary is then encoded into a URI using the 'postgresraster' provider metadata object.
After that the raster can be added to the project.

.. testcode:: loadlayer

 uri_config = {
     # database parameters
     'dbname':'gis_db',      # The PostgreSQL database to connect to.
     'host':'localhost',     # The host IP address or localhost.
     'port':'5432',          # The port to connect on.
     'sslmode':QgsDataSourceUri.SslDisable, # SslAllow, SslPrefer, SslRequire, SslVerifyCa, SslVerifyFull
     # user and password are not needed if stored in the authcfg or service
     'authcfg':'QconfigId',  # The QGIS athentication database ID holding connection details.

>      'service':NULL,         # The PostgreSQL service to be used for connection to the database.

Reviewer: ```suggestion
     'service': None,         # The PostgreSQL service to be used for connection to the database.
```
Travis is complaining about the NULL. Could you try with `None` for this one and the rest?

     'password':None,        # The PostgreSQL password for the user.
     # table and raster column details
     'schema':'public',      # The database schema that the table is located in.
     'table':'my_rasters',   # The database table to be loaded.
     'geometrycolumn':'rast',# raster column in PostGIS table
     'sql':None,             # An SQL WHERE clause. It should be placed at the end of the string.
     'key':None,             # A key column from the table.
     'srid':None,            # A string designating the SRID of the coordinate reference system.
     'estimatedmetadata':'False', # A boolean value telling if the metadata is estimated.
     'type':None,            # A WKT string designating the WKB Type.
     'selectatid':None,      # Set to True to disable selection by feature ID.
     'options':None,         # other PostgreSQL connection options not in this list.
     'enableTime': None,
     'temporalDefaultTime': None,
     'temporalFieldIndex': None,     
     'mode':'2',             # GDAL 'mode' parameter, 2 unions raster tiles, 1 adds tiles separately (may require user input) 
 }
 # remove any NULL parameters

>  uri_config = {key:val for key, val in uri_config.items() if val != NULL}

Reviewer: ```suggestion
 uri_config = {key:val for key, val in uri_config.items() if val is not None}
```
Overlooked this one?

 # get the metadata for the raster provider and configure the URI

>  md = QgsProviderRegistry.instance().providerMetadata('postgresraster')

Reviewer: Travis:
`NameError: name 'QgsProviderRegistry' is not defined`
Missing import?
And what about QgsDataSourceUri (line 349)?

This creates a new layer and adds it to the current project (making it appear
in the layer list) in one step.

To load a PostGIS raster:

PostGIS rasters, similar to PostGIS vectors, can be added to a project using a URI string.

> It is efficient to create a dictionary of strings for the database connection parameters.

Author: The code documentation for this has not been written yet, since it is very recent code. @elpaso may have another recommendation.

> It is efficient to create a dictionary of strings for the database connection parameters.

Reviewer 2: AFAIK none of the connection parameters are documented for any provider and they are not even part of the stable API. 

It would be nice to have though but if we want to get there it must be done in some automatic way so that we can query the provider to know what parameters it supports.

For now, you can either look at the provider's C++ implementation or look at layer.source() after a layer has been created in QGIS.

The dictionary is then loaded into an empty URI, before adding the raster.

> Note that None should be used when it is desired to leave the parameter blank: 

Author: This section was rewritten, and the reference to None removed.

> .. code-block:: python

Author: Done, but I need help with what the test should be at the end of the testcode block.

>  uri_config = {#

Author: Done.

>  # database parameters

Author: Done.

 'dbname':'gis_db', # The PostgreSQL database to connect to.
 'host':'localhost',     # The host IP address or localhost.
 'port':'5432',          # The port to connect on.
 # SSL/TLS mode, comment/uncomment for your pyQgis version

>  'sslmode':QgsDataSourceUri.SslDisable, # pyQgis 3.13+ SslAllow, SslDisable, SslPrefer, SslRequire, SslVerifyCa, SslVerifyFull

Author: Done.

>  #'sslmode':'disable',    # pyQgis 3.12 allow, disable, prefer, require, verify-ca, verify-full

Author: Done.

     'mode':'2',             # GDAL 'mode' parameter, 2 unions raster tiles, 1 adds tiles separately (may require user input) 
 }
 # remove any NULL parameters

>  uri_config = {key:val for key, val in uri_config.items() if val is not None}

Author: Yes I missed this. Thanks.
